### PR TITLE
add Container#stop

### LINKED ISSUE #90: How `stop` callback is invoked?
Consider following example from documentation

```
Application.finalize(:persistence) do |container|
  start do
    require 'sequel'
    container.register('persistence.db', Sequel.connect(ENV['DB_URL']))
  end

  stop do
    db.disconnect
  end
end
```

When `stop` callback (which calls `db.disconnect`) will be called? Is there a method to *stop* all components at once (like `finalize!` which *starts* all components)

This case is not covered in documentation...

### PR DESCRIPTION
**Pull request by GustavoCaso**
Fixes #90 

@solnic @mainameiz I have added the `stop` method in `Container`.

It also raises an error when trying to stop a component that has not started yet.

### CODE REVIEW

**Comment by timriley** [comment]
@GustavoCaso Rather than raise an error, couldn't this just gracefully skip any components not yet booted?

File: lib/dry/system/errors.rb [hunk 1]
```diff
@@ -58,5 +58,14 @@ def initialize(name)
         super("component identifier #{name.inspect} must be a symbol")
       end
     end
+
+    # Error raised when trying to stop a component that hasn't started yet
+    #
+    # @api public
+    ComponentNotStaredError = Class.new(StandardError) do
```
Comment (timriley) [inline lib/dry/system/errors.rb]: There's a typo here. It should be `ComponentNotStaredError`.

File: lib/dry/system/errors.rb [hunk 2]
```diff
@@ -58,5 +58,14 @@ def initialize(name)
...
+      def initialize(companent_name)
+        super("You can not stop a component that has not been started yet. Please use `start` method to start the component. system.start[:#{companent_name}]")
```
Comment (timriley) [inline lib/dry/system/errors.rb]: I feel like we could tighten this up a bit:

```ruby
"component +#{component_name}+ has not been started"
```

(I'm not sure we need the usage tip at this point)

File: lib/dry/system/errors.rb (hunk 1 above)
Comment (solnic) [reply lib/dry/system/errors.rb]: `ComponentNotStaredError` => `ComponentNotStartedError`

File: lib/dry/system/errors.rb (hunk 2 above)
Comment (solnic) [reply lib/dry/system/errors.rb]: `companent_name` => `component_name`

**Comment by mainameiz** [comment]
@GustavoCaso Thanks a lot!

File: lib/dry/system/errors.rb (hunk 1 above)
Comment (GustavoCaso) [reply lib/dry/system/errors.rb]: Bloody typo 🤣

File: lib/dry/system/errors.rb (hunk 2 above)
Comment (GustavoCaso) [reply lib/dry/system/errors.rb]: @timriley we can remove the tip, and I will fix the typo as well

**Comment by GustavoCaso** [comment]
@timriley @solnic I have fixed the typos and remove the tip from the error message. If everything looks good to you. I can merge and update the CHANGELOG and start working on the `shutdown` feature 😄

File: lib/dry/system/errors.rb [hunk 3]
```diff
@@ -58,5 +58,14 @@ def initialize(name)
         super("component identifier #{name.inspect} must be a symbol")
       end
     end
+
+    # Error raised when trying to stop a component that hasn't started yet
+    #
+    # @api public
+    ComponentNotStartedError = Class.new(StandardError) do
+      def initialize(component_name)
+        super("You can not stop a component +#{component_name}+ has not been started.")
```
Comment (timriley) [inline lib/dry/system/errors.rb]: @GustavoCaso Would you mind changing it to `"component +#{component_name}+ has not been started"`?

File: spec/integration/boot_spec.rb [hunk 4]
```diff
@@ -63,11 +63,62 @@ class Db < OpenStruct
...
+      end
+      system.start(:db)
+
+      conn = system['db.conn']
+      system.stop(:db)
+
+      expect(conn.established).to eq false
+    end
+
+    it 'raises an error when trying to stop a component that has not bee started' do
```
Comment (timriley) [inline spec/integration/boot_spec.rb]: typo, `bee` -> `been`

File: spec/integration/boot_spec.rb [hunk 5]
```diff
@@ -63,11 +63,62 @@ class Db < OpenStruct
...
+        stop do
+          db.conn.established = false
+        end
+      end
       conn = system['db.conn']
 
       expect(conn).to be_established
     end
+
+    it 'allows stop component' do
```
Comment (timriley) [inline spec/integration/boot_spec.rb]: This would read better as "allows component to be stopped" :)

**Comment by timriley** [comment]
@GustavoCaso Left a couple more wording notes for you, happy for you to merge after that 👍

File: lib/dry/system/errors.rb (hunk 3 above)
Comment (GustavoCaso) [reply lib/dry/system/errors.rb]: Sure thing. Sorry if I did not removed in the last commit, I understood only removing the tip part

Comment (GustavoCaso) [reply lib/dry/system/errors.rb]: @timriley done

**Pull request merged**